### PR TITLE
feature: integrates spandrel for upscaling

### PR DESCRIPTION
**Pull request by jaydrennan**

### CODE REVIEW

File: imaginairy/cli/upscale.py [hunk 1]
```diff
@@ -40,27 +31,15 @@ def validate_ext(ctx, param, value):
...
-    "--date",
-    callback=validate_date,
-    help="Specify the date format, used with --format.",
-    type=str,
-)
-@click.option(
-    "--ext",
-    callback=validate_ext,
-    help="Specify the file extension, used with --format.",
+    default="{original}_upscaled_{number}_{algorithm}_{date:%Y-%m-%d}_{time:%H-%M-%S}",
```
Comment (brycedrennan) [inline imaginairy/cli/upscale.py:40]: variable names are too vague.  The first time a person reads this they should know exactly what they are.

File: imaginairy/cli/upscale.py [hunk 2]
```diff
@@ -40,27 +31,15 @@ def validate_ext(ctx, param, value):
...
     type=str,
+    help="Format file name using {original},{number},{algorithm},{date}, and {time}.",
```
Comment (brycedrennan) [inline imaginairy/cli/upscale.py:42]: add explanation of what syntax is and how to do zero padding and date formatting

File: imaginairy/enhancers/upscale.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 
 def upscale_image(
     img: LazyLoadingImage | Image.Image,
-    upscaler_model: str = "realesrgan",
+    upscaler_model: str,
```
Comment (brycedrennan) [inline imaginairy/enhancers/upscale.py:31]: why was default removed

File: tests/test_utils/test_format_file_name.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        (
+            "{date}_{time}",
+            {"date": "2023/01/23", "time": "12-00", "ext": ".log"},
+            "2023/01/23_12-00.log",
+        ),
+        ("{original}", {"original": "file", "ext": ""}, "file"),
+    ],
+)
+def test_format_filename(format_str, data, expected):
+    if isinstance(expected, type) and issubclass(expected, Exception):
```
Comment (brycedrennan) [inline tests/test_utils/test_format_file_name.py:55]: Its bad practice to do generic exception catching.  The command line will display ValueErrors as messages to the user about what mistake they made.  So catch expected errors and the re-raise more descriptive useful ValueError.

**Comment by brycedrennan** [comment]
`aimg upscale tests/data/sand_upscale_difficult.jpg --upscale-model https://huggingface.co/city96/RevDiff/resolve/main/RevDiff1_x2_sharp.pth`
```
FileNotFoundError: [Errno 2] No such file or directory: './outputs/upscaled/sand_upscale_difficult_upscaled_5762_https://huggingface.co/city96/RevDiff/resolve/main/RevDiff1_x2_sharp.pth_2024-01-23_22-37-21.jpg'
```

File: imaginairy/cli/upscale.py (hunk 1 above)
Comment (brycedrennan) [reply imaginairy/cli/upscale.py:40]: If you were upscaling a file, is this the format you'd find most useful?

Comment (jaydrennan) [reply imaginairy/cli/upscale.py:40]: I've updated the variable names, joined date and time to be now, and updated the default structure to match something closer to how standard images are formatted.

File: imaginairy/cli/upscale.py (hunk 2 above)
Comment (jaydrennan) [reply imaginairy/cli/upscale.py:42]: Added zero padding and syntax explanations.

File: imaginairy/enhancers/upscale.py (hunk 3 above)
Comment (jaydrennan) [reply imaginairy/enhancers/upscale.py:31]: Will add it back. Thought it was confusing to have default set in two places.

Should it be "realesrgan"?

File: tests/test_utils/test_format_file_name.py (hunk 4 above)
Comment (jaydrennan) [reply tests/test_utils/test_format_file_name.py:55]: Will change the exception here.

**Comment by jaydrennan** [comment]
> `aimg upscale tests/data/sand_upscale_difficult.jpg --upscale-model https://huggingface.co/city96/RevDiff/resolve/main/RevDiff1_x2_sharp.pth`
> 
> ```
> FileNotFoundError: [Errno 2] No such file or directory: './outputs/upscaled/sand_upscale_difficult_upscaled_5762_https://huggingface.co/city96/RevDiff/resolve/main/RevDiff1_x2_sharp.pth_2024-01-23_22-37-21.jpg'
> ```

I'll change it so that urls can be submitted also.

File: imaginairy/enhancers/upscale.py (hunk 3 above)
Comment (jaydrennan) [reply imaginairy/enhancers/upscale.py:31]: I created "DEFAULT_UPSCALE_MODEL" in config.py and use that for the two applicable places.

File: imaginairy/cli/upscale.py (hunk 1 above)
Comment (jaydrennan) [reply imaginairy/cli/upscale.py:40]: For a default setting I wouldn't want the date and time as I don't think it's that useful.

I like the format the imaginiary generates file names, so it is now something closer to that along with a "UPSCALED" word added to it.

File: tests/test_utils/test_format_file_name.py (hunk 4 above)
Comment (jaydrennan) [reply tests/test_utils/test_format_file_name.py:55]: Separated exception handling to distinguish between expected and unexpected Exceptions.

**Comment by jaydrennan** [comment]
> > `aimg upscale tests/data/sand_upscale_difficult.jpg --upscale-model https://huggingface.co/city96/RevDiff/resolve/main/RevDiff1_x2_sharp.pth`
> > ```
> > FileNotFoundError: [Errno 2] No such file or directory: './outputs/upscaled/sand_upscale_difficult_upscaled_5762_https://huggingface.co/city96/RevDiff/resolve/main/RevDiff1_x2_sharp.pth_2024-01-23_22-37-21.jpg'
> > ```
> 
> I'll change it so that urls can be submitted also.

Changed so it pulls the file name from the url(without the extension) and assigns that as the name.

File: imaginairy/config.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 
 DEFAULT_MODEL_WEIGHTS = "sd15"
 DEFAULT_SOLVER = "ddim"
+DEFAULT_UPSCALE_MODEL = "realesrgan-x2-plus"
```
Comment (jaydrennan) [inline imaginairy/config.py:8]: What do you think the default should be?

File: imaginairy/api/upscale.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+from imaginairy.config import DEFAULT_UPSCALE_MODEL
+
+if TYPE_CHECKING:
+    from PIL import Image
+
+    from imaginairy.schema import LazyLoadingImage
+
+
+def upscale(
```
[the remark this answers is not in the document]
Comment (jaydrennan) [reply imaginairy/api/upscale.py]: Updated the name.

File: README.md [hunk 7]
```diff
@@ -349,14 +349,25 @@ When writing strength modifiers keep in mind that pixel values are between 0 and
...
-### Upscaling [by RealESRGAN](https://github.com/xinntao/Real-ESRGAN)
-```bash
->> imagine "colorful smoke" --steps 40 --upscale
-# upscale an existing image
->> aimg upscale my-image.jpg
-```
-<details>
-<summary>Python Example</summary>
+## Image Upscaling
+Upscale images easily using a vast array of models. Use "aimg upscale --list-models" to view built in models or submit
```
[the remark this answers is not in the document]
Comment (jaydrennan) [reply README.md]: Good point, I'll change this.

**Comment by brycedrennan** [comment]
Whats the status of this pull request?

**Review by brycedrennan — APPROVED** [review]

**Pull request merged**